coucharchive

```diff
@@ -307,10 +307,22 @@ class ReplicationControl(object):
         # Choose the value to achieve replication in ideal_duration:
         ideal_number = self._ideal_number_of_replications_for_ideal_duration()
 
-        # ... but if there were errors, take 0.9 × the lowest error value
-        if self._last_errors:
-            min_number_with_error = min(e[1] for e in self._last_errors)
-            ideal_number = int(0.9 * min_number_with_error)
+        # ... but if there were errors, compute a age weighted average
+        # between past erros and ideal_number.
+        now = time.time()
+        weights = []
+        weighted_values = []
+        target = [(now, ideal_number)]
+        for i in self._last_errors + target:
+            age = i[0] - now
+            # Weight for an old event is 0, weight for a recent event is 1,
+            # between the two it is a linear curve.
+            # 0.0033 = 1 / (5 * 60)
+            weight = max(0, 0.0033 * age + 1)
+            weights.append(weight)
+            # Value is multiply by 0.9 so we stay just below
+            weighted_values.append(weight * i[1] * 0.9)
+        ideal_number = sum(weighted_values) / sum(weights)
```

Comment:
- It seems that `weights` is useless and you could use:
  ```suggestion
          ideal_number = sum(weighted_values) / len(weighted_values)
  ```
- For a better understanding, what you put the 0.9 factor a the end, so that `weighted_values` are real weighted values?
  ```suggestion
          ideal_number = 0.9 * sum(weighted_values) / sum(weights)
  ```

Reply:
> It seems that weights is useless and you could use:

In weighted average, it's the sum of the weight that is used as the divider, not the number of values (like in a classic average computation).
See here https://en.wikipedia.org/wiki/Weighted_arithmetic_mean#Mathematical_definition
So I think your proposal would be wrong no?

> For a better understanding, what you put the 0.9 factor a the end, so that weighted_values are real weighted values?

If I do that, it would be applied to the `target = [(now, ideal_number)]` too, and ... I did not wanted that ... but I just release my code do the same ...
OK for you if I split the part that does the `* 0.9` in another loop that applied only to values and not to the target?

Reply:
> So I think your proposal would be wrong no?

You're totally right, please ignore my (dumb) remark! :smile: 

> OK for you if I split the part that does the * 0.9 in another loop that applied only to values and not to the target?

I'm not sure I understand what you mean? In any case, I think current code is *equivalent* to my proposal of `ideal_number = 0.9 * ...`.

Reply:
> You're totally right, please ignore my (dumb) remark! smile

No issue, I had to read wikipedia to remember that!

> > OK for you if I split the part that does the * 0.9 in another loop that applied only to values and not to the target?
> 
> I'm not sure I understand what you mean? In any case, I think current code is _equivalent_ to my proposal of `ideal_number = 0.9 * ...`.

Sorry I was not clear. Your proposal is indeed equivalent, 1 for 1, to the current code.
But the current code have a bug, in that it multiply the `target` per 128, so for example, if the target is the maximum, `128`, we will return at best `128 * 0.9` as the ideal number of worker.

Reply:
Discussion resolved by using a `last_errors_lowered` array:
```python
        target = [(now, ideal_number)]
        # Lower last errors values a bit (× 0.9) so we stay just below the
        # "error zone":
        last_errors_lowered = [(i[0], i[1] * 0.9) for i in self._last_errors]
        for i in last_errors_lowered + target:
            age = now - i[0]
```
thanks!